### ISSUE TITLE
Get wrong response when POST resource with reference by using Bundle Endpoint

### ISSUE THREAD

**Issue by rhinozD** [awslabs/fhir-works-on-aws-deployment#305]
Hi FHIR Works Team,
I tried to use Bundle Endpoint to POST Observation resource with subject and device reference.
Here was my request body:
```
{
    "resourceType": "Bundle",
    "type": "transaction",
    "entry": [
        {
            "resource": {
                "resourceType": "Observation",
                "identifier": [
                    {
                        "value": "dfhjshdfjshfk754845789"
                    }
                ],
                "status": "final",
                "category": [
                    {
                        "coding": [
                            {
                                "version": "v1",
                                "code": "ABC",
                                "display": "display value for ABC"
                            }
                        ]
                    }
                ],
                "code": {
                    "coding": [
                        {
                            "code": "LQA",
                            "display": "display value for LQA"
                        }
                    ]
                },
                "subject": {
                    "reference": "Patient/1324efefg-e638-4817-8692-sfdfs"
                },
                "device": {
                    "reference": "Device/3243ffd-c2c4-473e-9a57-gsgsdg"
                },
                "effectiveDateTime": "2021-01-09T20:00:06Z",
                "valueQuantity": {
                    "unit": "cm",
                    "value": 170
                },
                "note": [
                    {
                        "text": "This is a note"
                    }
                ]
            },
            "request": {
                "method": "POST",
                "url": "Observation"
            }
        }
    ]
}
```
I got an error: `"Cannot read property 'length' of undefined"`.
When I changed subject and device to: 
```
                "subject": {
                    "reference": "{{API_URL}}/Patient/1324efefg-e638-4817-8692-sfdfs"
                },
                "device": {
                    "reference": "{{API_URL}}/Device/3243ffd-c2c4-473e-9a57-gsgsdg"
                }
```
it worked; I could POST Observation using Bundle Endpoint.
BUT when I make a GET Observation request by the id of Observation resource that was POSTed successfully above.
it returned:
```
    "subject": {
        "reference": "Observation/1324efefg-e638-4817-8692-sfdfs"
    },
    "device": {
        "reference": "Observation/3243ffd-c2c4-473e-9a57-gsgsdg"
    }
```

is this a bug? 

Thanks in advance for any help you are able to provide.

**Comment by carvantes** [comment]
Hi,

Thanks for the detailed bug report. It looks like there are 2 separate issues:

first, your Bundle entry is missing a `fullUrl`. The use of relative references in a Bundle entry without a fullUrl is a bit of a gray area in the FHIR Spec. https://www.hl7.org/implement/standards/fhir/bundle.html#references

Validation using the [FHIR Validator](https://www.hl7.org/fhir/validation.html#jar) outright rejects it:
```
Error @ Bundle.entry[0].resource.ofType(Observation).subject : Relative Reference appears inside Bundle whose entry is missing a fullUrl
```

Adding `fullUrl` to the Bundle entry will successfully execute the transaction.
```
...
    "entry": [
        {
           “fullUrl”: “urn:uuid:88f151c0-a954-468a-88bd-5ae15c08e059”, // not necessarily a uuid. any string will do
            "resource": {
                "resourceType": "Observation",
...
```
However, popular implementations such as HAPI FHIR accept the use case of missing fullUrl + relative references, and simply assume that relative references are relative to the base URL of the server. I think that many customers(myself included) intuitively expect this behavior.

We'll go ahead and make a fix to support this use case.

---

second,  the issue with absolute URL references being replaced is a bug.
(`"reference": "{{API_URL}}/Patient/1324efefg-e638-4817-8692-sfdfs"` being replaced by `"reference": "Observation/1324efefg-e638-4817-8692-sfdfs"`)

We were able to repro the bug and we'll start working on a fix.

**Comment by carvantes** [comment]
Hi,

Both fixes have been released as part of `fhir-works-on-aws-routing@5.2.1`

Thanks again for reporting the issue.
